### ISSUE TITLE
same error eccountered as #2 , but cannot fix with cudnn 8 and cuda 10.2

### ISSUE THREAD

**Issue by Mohan-Zhang-u** [idiap/linear-transformer-experiments#5]
same as #2 : I tried pytorch version 1.9.1 and 1.8.2 LTS, cuda 10.2 and 11.2, but none of them could solve the issue `TypeError: 'NoneType' object is not callable` when performing `CausalDotProduct.dot[device.type]`

**Comment by angeloskath** [comment]
Hi,

The error means that the `fast-transformers` library is not properly installed. The simplest way to debug this would be to `git clone` the `fast-transformers` repository and run `python setup.py build_ext --inplace` to see if it properly builds the extensions. Afterwards you can also `pip install -e .` to install in development mode from within the repository.

Let me know if you encounter more errors.

Cheers,
Angelos

**Comment by Mohan-Zhang-u** [comment]
Just did as instructed, but the same `'NoneType' object is not callable` error appears at the same location. Let me also try the solution proposed [here](https://github.com/idiap/linear-transformer-experiments/issues/2#issuecomment-926521451)

**Comment by angeloskath** [comment]
Hm, but did the compilation finish successfully? Can you see the compiled shared object binary file?

**Comment by Mohan-Zhang-u** [comment]
> Hm, but did the compilation finish successfully? Can you see the compiled shared object binary file?

yes the compilation finished successfully. Is there anything that I could share to help you address the issue?

**Comment by angeloskath** [comment]
So, the error means that the `causal_product_cuda` module is not found. This can be either that it is not compiled or that it is compiled for a different python or that it is not compiled for CUDA. So if you make sure that `nvcc` is available I would advise to make a new virtual environment and install `pytorch-fast-transformers` in that new environment. Check that the CUDA extensions are properly compiled and then use that repo.

In any way, I would advise you to browse the `pytorch-fast-transformers` repository issues for similar situations. Most of the time it is an improperly install CUDA environment.

Cheers,
Angelos

**Comment by Mohan-Zhang-u** [comment]
> So, the error means that the `causal_product_cuda` module is not found. This can be either that it is not compiled or that it is compiled for a different python or that it is not compiled for CUDA. So if you make sure that `nvcc` is available I would advise to make a new virtual environment and install `pytorch-fast-transformers` in that new environment. Check that the CUDA extensions are properly compiled and then use that repo.
> 
> In any way, I would advise you to browse the `pytorch-fast-transformers` repository issues for similar situations. Most of the time it is an improperly install CUDA environment.
> 
> Cheers,
> Angelos

Sure and thanks for the suggestion! I did start from a new virtual env every time I performed the installation and nvcc was there. I will recheck my CUDA and cuDNN settings and see whether I can find anything wrong. Thanks a lot for your help!
